filter.go

```diff
@@ -43,31 +43,66 @@ const (
 	MapResultSkipDir
 )
 
-func Walk(ctx context.Context, p string, opt *WalkOpt, fn filepath.WalkFunc) error {
-	root, err := filepath.EvalSymlinks(p)
-	if err != nil {
-		return errors.WithStack(&os.PathError{Op: "resolve", Path: root, Err: err})
-	}
-	rootFI, err := os.Stat(root)
-	if err != nil {
-		return errors.WithStack(err)
+type filterFS struct {
+	fs  FS
+	opt *FilterOpt
+}
+
+func NewFilterFS(fs FS, opt *FilterOpt) FS {
+	if opt == nil {
+		return fs
 	}
-	if !rootFI.IsDir() {
-		return errors.WithStack(&os.PathError{Op: "walk", Path: root, Err: syscall.ENOTDIR})
+
+	if fs, ok := fs.(*filterFS); ok {
```

Comment:
Note to self: need to add some tests to check that this kind of combining is actually valid in all cases.

Reply:
Still need to add some tests, but, I've convinced myself this should be good.

IncludePatterns and ExcludePatterns both set `skip = true` on a match, so it's a purely additive check, so the order really shouldn't matter. The paths stay consistent between Walks, so again, just doing the naive concatenation should be fine.

Reply:
This definitely needs tests or if there isn't a performance issue then it can be removed in the initial version. Otherwise, we have 2 competing interfaces that can easily get out of sync.

Reply:
This actually isn't consistent, so we need to remove this additional check. See the new test, it's not valid to combine inclusions/exclusions as shown.